### DISCUSSION TITLE
How to reuse middleware? Confusion about "next" type.

### DISCUSSION THREAD

**Discussion by tkgalk** [go-kit/kit#1220] [Q&A]
Hello.

I'm wondering how you could go about reusing some of the middleware I've already written. In all of the examples I've seen, the middleware has a `next` field, with a next service hardcoded.

```
type loggingMiddleware struct {
	logger log.Logger
	next   SomeService
}
```

Since Go 1.18 was released, I'm thinking how I could make this more generic so that "any" service can be accepted instead of just the one. Would I really need to implement an authentication middleware for each of my services separately? I must be missing something. Thanks.

**Comment by mcosta74** [comment]
I'd put and `authenticationMiddleware` at endpoint level and not at service level, in this case it will be service independent

```go
func AuthMiddleware(next endpoint.Endpoint) endpoint.Endpoint {
    return func(ctx context.Context, request interface{}) (response interface{}, err error) {
         authHeader := ctx.Value(.....). // get the authentication information from the context

         authInfo, err := validate(authHeader)      // here is the authentication logic
         if err != nil { 
             return nil,  err
         }
         ctx = context.With(ctx, <someKey>, authInfo)
         return next(ctx, request)
    }
}
```

Service level middlewares are designed to enrich each method without changing the business logic (for instance to add logging, metrics, ...); to do that you need to know exactly which method is going to be executed
